app/views.py

```diff
@@ -0,0 +1,443 @@
+            )
+            .order_by(Post.timestamp.desc())
+            .slice(pages["first_post_index"], (pages["last_post_index"] + 1))
+        )
+    else:
+        posts = Post.query.order_by(Post.timestamp.desc()).slice(
+            pages["first_post_index"], (pages["last_post_index"] + 1)
+        )
+
+    num_posts = posts.count()
+    return render_template(
+        "home.html",
+        posts=posts,
+        page=page,
+        items_per_page=pages["items_per_page"],
+        num_posts=num_posts,
+    )
+
+
+@main.route("/users/<username>")
+def user_details(username):
+    try:
+        user = User.query.filter_by(username=username).one()
+    except exc.SQLAlchemyError:
+        abort(404)
+    match = False
+    if current_user == user:
+        match = True
+    posts = (
+        Post.query.filter_by(user_id=user.id).order_by(Post.timestamp.desc()).limit(1)
+    )
+    follow_count = Follow.query.filter_by(user_followed=user.id).count()
+    following_count = Follow.query.filter_by(user_following=user.id).count()
+    liked_count = Like.query.filter_by(user_id=user.id).count()
+    return render_template(
+        "user_details.html",
+        user=user,
+        match=match,
+        posts=posts,
+        follow_count=follow_count,
+        following_count=following_count,
+        liked_count=liked_count
+    )
+
+
+@main.route("/users/<username>/followers")
+def user_followers(username):
+    try:
+        user = User.query.filter_by(username=username).one()
+    except exc.SQLAlchemyError:
+        abort(404)
+    match = False
+    if current_user == user:
+        match = True
+    followers = user.followers
+    follow_count = Follow.query.filter_by(user_followed=user.id).count()
+    following_count = Follow.query.filter_by(user_following=user.id).count()
+    liked_count = Like.query.filter_by(user_id=user.id).count()
+    return render_template(
+        "user_followers.html",
+        user=user,
+        match=match,
+        followers=followers,
+        follow_count=follow_count,
+        following_count=following_count,
+        liked_count=liked_count
+    )
+
+
+@main.route("/users/<username>/following")
+def user_following(username):
+    try:
+        user = User.query.filter_by(username=username).one()
+    except exc.SQLAlchemyError:
+        abort(404)
+    match = False
+    if current_user == user:
+        match = True
+    following = user.following
+    follow_count = Follow.query.filter_by(user_followed=user.id).count()
+    following_count = Follow.query.filter_by(user_following=user.id).count()
+    liked_count = Like.query.filter_by(user_id=user.id).count()
+    return render_template(
+        "user_following.html",
+        user=user,
+        match=match,
+        following=following,
+        follow_count=follow_count,
+        following_count=following_count,
+        liked_count=liked_count
+    )
+
+
+@main.route("/users/<username>/likes")
+def user_likes(username):
+    try:
```

Comment:
Remove duplication here, we should extract a function that returns match and user